src/decode/vec2d.rs

```diff
@@ -0,0 +1,74 @@
+use std::ops::{Index, IndexMut};
+
+/// A 2 dimensional matrix in row-major order backed by a contiguous `Vec`
+#[derive(Debug)]
+pub struct Vec2D<T> {
+    data: Vec<T>,
+    cols: usize,
+    rows: usize,
+}
+
+impl<T> Vec2D<T> {
+    /// Initialize a grid of size `rows * cols` with the given data element.
+    pub fn init(rows: usize, cols: usize, data: T) -> Vec2D<T>
+    where
+        T: Clone,
+    {
+        Vec2D {
+            data: vec![data; rows * cols],
+            cols,
+            rows,
+        }
+    }
+
+    /// Fills the grid with elements by cloning `value`.
+    pub fn fill(&mut self, value: T)
+    where
+        T: Clone,
+    {
+        self.data.fill(value)
+    }
+}
+
+impl<T> Index<usize> for Vec2D<T> {
+    type Output = [T];
+
+    fn index(&self, idx: usize) -> &Self::Output {
+        if idx < self.rows {
+            let start_idx = idx * self.cols;
+            &self.data[start_idx..start_idx + self.cols]
+        } else {
+            panic!("row index {:?} out of bounds.", idx);
+        }
+    }
+}
+
+impl<T> IndexMut<usize> for Vec2D<T> {
+    fn index_mut(&mut self, idx: usize) -> &mut Self::Output {
+        &mut self.data[(idx * self.cols)..]
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
```

Comment:
Can you add unit tests for:
- `init()`,
- `index_mut()`,
- some `should_panic` in case of indexing, e.g. `vec2d[0][3]` should manic on a 2x3 matrix.

Reply:
added unit tests for a variety of edge cases for `IndexMut` and `Index`